pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -745,13 +745,50 @@ func (p *PriorityQueue) AssignedPodUpdated(pod *v1.Pod) {
 // if Pop() is waiting for an item, it receives the signal after all the pods are in the
 // queue and the head is the highest priority pod.
 func (p *PriorityQueue) moveAllToActiveOrBackoffQueue(event framework.ClusterEvent, preCheck PreEnqueueCheck) {
-	unschedulablePods := make([]*framework.QueuedPodInfo, 0, len(p.unschedulablePods.podInfoMap))
+	activated := false
+
 	for _, pInfo := range p.unschedulablePods.podInfoMap {
-		if preCheck == nil || preCheck(pInfo.Pod) {
-			unschedulablePods = append(unschedulablePods, pInfo)
+		isSchedulable := framework.PodMaybeSchedulable
+		if preCheck != nil {
+			isSchedulable = preCheck(pInfo.Pod)
+		}
+
+		if isSchedulable == framework.PodNotAffected {
+			continue
+		}
+
+		// If the event doesn't help making the Pod schedulable, then keep it in its current state.
+		// Note: we don't run the check if pInfo.UnschedulablePlugins is nil, which denotes
+		// either there is some abnormal error, or scheduling the pod failed by plugins other than PreFilter, Filter and Permit.
+		// In that case, it's desired to move it anyways.
+		if len(pInfo.UnschedulablePlugins) != 0 && !p.podMatchesEvent(pInfo, event) {
+			continue
+		}
+
+		pod := pInfo.Pod
+		if isSchedulable != framework.PodImmediatelySchedulable && p.isPodBackingoff(pInfo) {
```

Comment:
any chance to de-dupe the code with movePodsToActiveOrBackoffQueue(), or we need a follow-up to revamp the code.

Reply:
Done. I moved the common code for moving to active or backoff queue into a new `movePodToActiveOrBackoffQueue` and kept the different logic for picking pods and checking the filter in `moveAllToActiveOrBackoffQueue` and `movePodsToActiveOrBackoffQueue`.

I also restored setting `p.moveRequestCycle` in `movePodsToActiveOrBackoffQueue`, like it was before. Whether the cluster events for which that function gets called (`UnschedulableTimeout` in `flushUnschedulablePodsLeftover`, `AssignedPodAdd` in `AssignedPodAdded`, `AssignedPodUpdate` in `AssignedPodUpdated`) really need that is something that we can discuss later when improving the handling of false positives for concurrent cluster events.